python/ray/air/_internal/remote_storage.py

```diff
@@ -94,10 +94,16 @@ def get_fs_and_path(
         fs = _cached_fs[cache_key]
         return fs, path
 
+    # In case of hdfs filesystem, if uri does not have the netloc part below will
+    # failed with hdfs access error.  For example 'hdfs:///user_folder/...' will
+    # fail, while only 'hdfs://namenode_server/user_foler/...' will work
+    if parsed.scheme == "hdfs" and parsed.netloc == "" and not fsspec:
+        return None, None
     try:
-        fs, path = pyarrow.fs.FileSystem.from_uri(uri)
-        _cached_fs[cache_key] = fs
-        return fs, path
+        if not (parsed.scheme == "hdfs" and parsed.netloc == ""):
```

Comment:
Why this check is still needed given the early return above?

Reply:
for the case of (parsed.scheme == "hdfs" and parsed.netloc == "" and fsspec), it will not enter above if clause, and not return (None, None), and it will come into the try clause.  Without if at line 103, it will try to access pyarrow with parsed.netloc == "" and pyarrow will throw error.  Remember we used to discuss, we don't want pyarrow to run into any issue when first time initialize, because that will make pyarrow in a broken state and even later with correct file path, pyarrow will not be able to recover and will continue throw error.

So the strategy here is to prevent any problematic initialization for pyarrow at early stage from the very beginning.

Reply:
Ok, can they be combined them together, so we can localize/contain this special handling (instead of spreading around), something like:
```
try:
    # COMMENT HERE
    if parsed.scheme == "hdfs" and parsed.netloc == "":
        if not fsspec:
            return None, None
    else:
        fs, path = pyarrow.fs.FileSystem.from_uri(uri)
                _cached_fs[cache_key] = fs
                return fs, path
except ():
  ....
....
```

Reply:
will try